### PR TITLE
Update and modernize dependencies and examples

### PR DESCRIPTION
**Pull request by kiview**
This PR shall act as a starting point for updating the workshop. For now, this updates Java and most dependencies (with updating Testcontainers being the most important one).

We also removed Maven for now, since the last part of the workshop (out-of-process testing) currently did only support Gradle. However, we can consider removing this step altogether and re-instating Maven.

Also, the instructions were slightly tweaked, to make it easier to follow the workshop by yourself.

### CODE REVIEW

File: step-4-your-first-testcontainers-integration.md [hunk 1]
```diff
@@ -10,38 +10,39 @@ It means that starting to use Testcontainers in our project \(once we add a depe
...
-        ℹ︎ Checking the system...
-        ✔ Docker version should be at least 1.6.0
-        ✔ Docker environment should have more than 2GB free disk space
-        ✔ File should be mountable
+                   
+    12:26:58.658  INFO 32972 --- [    Test worker] o.t.d.DockerClientProviderStrategy       : Found Docker environment with Environment variables, system properties and defaults. Resolved dockerHost=tcp://127.0.0.1:56956
+    12:26:58.660  INFO 32972 --- [    Test worker] org.testcontainers.DockerClientFactory   : Docker host IP address is 127.0.0.1
+    12:26:58.784  INFO 32972 --- [    Test worker] org.testcontainers.DockerClientFactory   : Connected to docker:
+      Server Version: 66+testcontainerscloud
+      API Version: 1.41t com.example.demo.DemoApplicationTest
```
Comment (bsideup) [inline step-4-your-first-testcontainers-integration.md:33]: huh? :D

File: step-6-adding-redis.md [hunk 2]
```diff
@@ -1,30 +1,31 @@
...
+static final GenericContainer redis = new GenericContainer("redis:6-alpine")
+                                            .withExposedPorts(6379);
+
+static {
+  redis.start();
+}
+
+@DynamicPropertySource
+public static void configureRedis(DynamicPropertyRegistry registry) {
+  registry.add("spring.redis.host", redis::getHost);
```
Comment (bsideup) [inline step-6-adding-redis.md:23]:
```suggestion
@DynamicPropertySource
public static void configureRedis(DynamicPropertyRegistry registry) {
  redis.start();
  registry.add("spring.redis.host", redis::getHost);
```

File: step-8-edge-cases.md [hunk 3]
```diff
@@ -50,22 +52,54 @@ But since we're not using Spring Context here, we need to create an instance of
...
-If we run the tests, we discover that there is a limit of 64bit number for the increment, oh well!
+The test should fail with a somewhat cryptic error:
+```text
+Error in execution; nested exception is io.lettuce.core.RedisCommandExecutionException: ERR increment or decrement would overflow
+```
+And there's nothing on your side to fix, the test is pushing the boundaries of Redis.  
+But on the bright side we learned how to use Testcontainers outside of the Spring Framework. 
+We also saw how we can utilize to learn about the limitations and behavior of extra components.
+
+Delete the test before anyone notices. Just kidding, let's turn this into a useful test by asserting we get the exceptions.
```
Comment (bsideup) [inline step-8-edge-cases.md:84]: IMO we shouldn't be asserting faulty behaviour (nobody would expect a request to fail on production, so why would test do?)

File: step-6-adding-redis.md (hunk 2 above)
Comment (kiview) [reply step-6-adding-redis.md:23]: I was never sure if this would change semantics as opposed to our singleton approach, if users would have something like `@DirtiesContext(classMode = DirtiesContext.ClassMode.AFTER_EACH_TEST_METHOD)` and recreate the context.

So I tried it out and indeed, the containers are only started a single time per test-suite, very nice!

File: step-8-edge-cases.md (hunk 3 above)
Comment (kiview) [reply step-8-edge-cases.md:84]: The alternative (I think we discussed with @rnorth) would be to change the production code to validate the input and consciously throw an `InvalidArgumentException`, which we could then assert here.

File: step-4-your-first-testcontainers-integration.md (hunk 1 above)
Comment (kiview) [reply step-4-your-first-testcontainers-integration.md]: Replace with Docker Desktop output.

File: step-3-adding-some-tests.md [hunk 4]
```diff
@@ -41,7 +45,7 @@ CREATE TABLE IF NOT EXISTS talks(
 
 INSERT
   INTO talks (id, title)
-  VALUES ('welcome-to-junit-5', 'JUnit 5.2 actually :)')
+  VALUES ('welcome-to-junit-5', 'JUnit 5.x actually :)')
```
Comment (bsideup) [inline step-3-adding-some-tests.md:48]: FYI the reason why it was `5.2` is because that was a reference to a talk from Spring I/O, and, with `5.x`, and makes very little sense - perhaps we should use something less weird then? :)

File: step-8-edge-cases.md (hunk 3 above)
Comment (bsideup) [reply step-8-edge-cases.md:84]: That's one of the reasons why this step never had a solution offered - the test shouldn't be changed (this would be a really bad thing to recommend doing, I think we all agree here), and it is unclear how to change the implementation (validate? Catch and ignore the overflow? Something else?)

Anyways, if we want to provide a solution, I suggest we change the implementation and not the test. After all, this isn't test's fault :D

File: step-3-adding-some-tests.md (hunk 4 above)
Comment (kiview) [reply step-3-adding-some-tests.md:48]: shall we just put a Testcontainers talk in there? :) 
`Modern Integration Testing with Testcontainers`

Comment (bsideup) [reply step-3-adding-some-tests.md:48]: Yep, let's change 👍

File: step-8-edge-cases.md (hunk 3 above)
Comment (kiview) [reply step-8-edge-cases.md:84]: IMO the test helped us in exploring the behavior of the product we are using. With the behavior being discovered through the test, it is IMO very valid to change the implementation and maybe even the test itself to incorporate the change in design (this is an interactive feedback loop, in the style of TDD). 

So I think it would make sense to change the prod code to throw a meaningful custom exception (`MaxRatingsAddedException`?) and assert this in the test. I don't think the repository can make any meaningful business decisions about what else to do (catch and ignore feels weird).

Comment (bsideup) [reply step-8-edge-cases.md:84]: Okay, as a compromise, `MaxRatingsAddedException` would be fine (and pull the decision on how to handle it on the upper level).

I am still not in favour of changing a test here (because a good test should reflect how system should work, not how it is implemented), but `MaxRatingsAddedException` would at least remove implementation-specific details (`RedisSystemException` is cryptic)

Comment (kiview) [reply step-8-edge-cases.md:84]: Sounds good 👍 

> I am still not in favour of changing a test here (because a good test should reflect how system should work, not how it is implemented), but MaxRatingsAddedException would at least remove implementation-specific details

Generally yes, but sometimes you learn about the limitations of the system only during implementation and then you might need to make new decisions on how the business logic needs to handle this (or if this limitation is mostly a theoretical one).

Whatever, I think this discussion is somewhat academic and not that important for this small example, but I think it's great we have it. I will adapt it accordingly in the docs 👍

**Review by bsideup — APPROVED** [review]
